File: Android/app/src/main/java/com/majinnaibu/monstercards/models/SearchResultItem.java
package com.majinnaibu.monstercards.models;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class SearchResultItem {
    public enum Type {
        MONSTER,
        COLLECTION
    }

    @NonNull
    public final Type type;

    @Nullable
    public final Monster monster;

    @Nullable
    public final Collection collection;

    public SearchResultItem(@NonNull Monster monster) {
        this.type = Type.MONSTER;
        this.monster = monster;
        this.collection = null;
    }

    public SearchResultItem(@NonNull Collection collection) {
        this.type = Type.COLLECTION;
        this.monster = null;
        this.collection = collection;
    }
}

File: iOS/MonsterCards/Models/SearchResultItem.swift
//
//  SearchResultItem.swift
//  MonsterCards
//
//  Ported from Android SearchResultItem.java.
//  Disambiguation type for search-result items across collections and dashboards.

import Foundation

/// Discriminant for a SearchResultItem's concrete entity type.
public enum SearchResultItemType: String, Codable, CaseIterable {
    case monster
    case collection
}

/// Lightweight non-persisted wrapper around either a Monster or a Collection,
/// with enough metadata to distinguish cross-collection search results.
public struct SearchResultItem: Identifiable, Hashable {

    public let id: String         // objectID-based unique key.
    public let type: SearchResultItemType
    public let monster: Monster?
    public let collection: Collection?

    // MARK: - Constructors mirroring Android

    /// Wraps a Monster — sets `type = .monster`.
    public init(monster: Monster) {
        self.id = monster.objectID.uriRepresentation().lastPathComponent
        self.type = .monster
        self.monster = monster
        self.collection = nil
    }

    /// Wraps a Collection — sets `type = .collection`.
    public init(collection: Collection) {
        self.id = collection.objectID.uriRepresentation().lastPathComponent
        self.type = .collection
        self.monster = nil
        self.collection = collection
    }

    // MARK: - Identifiable / Hashable

    public static func == (lhs: SearchResultItem, rhs: SearchResultItem) -> Bool {
        lhs.id == rhs.id && lhs.type == rhs.type
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(id)
        hasher.combine(type)
    }
}
